scripts/validate_docstrings.py

```diff
@@ -691,7 +691,11 @@ def get_validation_data(doc):
 
     if doc.is_function_or_method:
         if not doc.returns:
-            if 'return' in doc.method_source:
+            if re.search(r""" \"\"\".*\"\"\".*  # Skip the docstring.
```

Comment:
@WillAyd To your first question: I don't believe so. The if statement reads True if the search finds at least one match. The return on the second line can match with any return command on the function, and thus the search will consider all this possibilities.

For example, on the following function:
```
def func():
  """
  docstring
  """
  if something:
    return None
  else:
    return 7
```

The regex can evaluate the first return, but it is also viable for it to skip the docstring and then, with the `.*` at the end of the first line, consume the block
```
 if something:
    return None
  else:
```
leaving the return on the regex's second line to match the 'return 7'. Thus, the regex considers all returns on the function.

Reply:
Do you see adequate test coverage for that? Would be nice to add.

Still would prefer something closer to the finditer approach described above as using DOTALL and maybe MULTILINE in coordination with greedy operators is tougher to reason about

Reply:
About using the find function instead of search: The find function looks for all non-overlapping matches, while the search returns only the first one, which is sufficient and bound to have a better performance.

Unless you are suggesting that I use find to get all occurrences of return and then evaluate if it's None or bare outside the regex, but I think that would just add more complexity to the code.

Reply:
About `return(?!\n| None)`, that would fail if I had a comment line ending with "return" plus some other word.

The purpose of the regex's first line and the `\n\s*` on the second is to ensure that the word "return" is not in a comment or in the docstring. By including `\n\s*`, the regex will only accept returns that are on the beginning of a line, save for identation, what guarantees they are a actual command.

Then, on the third line, the one that ensures that the return isn't bare nor None, I couldn't check only `(?!\n| None)`, because that:

- would not exclude `return  #some comment`, creating a false positive
- would not accept `return None if something.empty() else some_operation(something)`, creating a false negative.

Also, the reason that I read 0 or plus whitespaces (`\s*`) on the third line is to prevent trailing whitespaces to have any impact. Consider the example:

Let's suppose I'm using something like `return(?!\n|  #| None\n)`, which is your suggestion plus the necessary to respect the 2 cases that I previously mentioned, and read `return None\n`, that would match the negative lookup (`(?!)`) and not match the whole regex, what would be the desired behavior. But if, instead, I read `return None \n`, it would fail. 

That whitespace can be a reminiscent of a previously existing but now removed comment on that line, for example. It could also be two whitespaces, or n, for all that I know. The same applies for to the previous `\s*` on the line, since I can have something like `return      None\n`, which is viable according to Python syntax.

That said, I know that there may be other validations (like PEP8's) that prevent absurd cases like 'return(3 whitespaces)None(4 whitespaces)\n', but I have no knowledge of them or where exactly they draw the line. Also, if one of those cases were to happen, it wouldn't be the place of this test to catch it, as the error message could be misleading.

Reply:
@WillAyd let me hear your thoughts on this.

Reply:
Sticking with my previous comment this can be greatly simplified. As mentioned I provided that pattern without testing or guaranteeing it's accuracy but I can't imagine the solution would be that far off. Shouldn't need to complicate things with DOTALL and MULTILINE flags here

Reply:
@WillAyd I had not seen your previous comment.

What do you mean by adequate test coverage? Writing another test function? I can do that.